### PR TITLE
New EMC VPLEX module to create storage-volume.

### PR DESCRIPTION
**Pull request by hwakabh**
##### SUMMARY

New modules to manage storage-volume on EMC VPLEX.  
Allows creating storage-volume encapsulation from backend storage-array.  

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

storage/emc/emc_vplex_storagevolume  

##### ADDITIONAL INFORMATION

- VPLEX storage-volume encapsulation should meet the requirements below:
  - Some storage-array supported are properly exposed to VPLEX with backend zoning
  - LUN(s) are already created on backend storage-array and exported to VPLEX

### CODE REVIEW

**Comment by remixtj** [comment]
I've looked through the module and i see several usage of logging. I don't know if it is a good practice, and anyway i'd prefer to have logging set as optional. Adding some notes now to the code.

File: lib/ansible/modules/storage/emc/emc_vplex_storagevolume.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+    vplex = VPLEX(
+        ip_address=module.params["vplex_ip_address"],
+        username=module.params["vplex_username"],
+        password=module.params["vplex_password"])
+
+    # --- if S/N differs from expected, exit the program
+    if not vplex.confirm_vplex_serial_number(expect_serial_number=module.params["vplex_serialnum"]):
+        logger.error('Target system S/N and expected ones are different. Exit the module wihtout any operations.')
+        module.exit_json(changed=False)
```
Comment (remixtj) [inline lib/ansible/modules/storage/emc/emc_vplex_storagevolume.py]: When hitting an error you must exit with module.fail_json

File: lib/ansible/modules/storage/emc/emc_vplex_storagevolume.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    for storage_volume in storage_volumes_list_all['response']['context'][0]['children']:
+        storage_volumes.append(storage_volume['name'])
+
+    if vpd_id not in storage_volumes:
+        logger.info('------- Execute array Re-discovery')
+        try:
+            rediscover_storage_array(vplex=vplex, array_name=array_name)
+        except Exception:
+            logger.error('Failed to array re-discover...')
+            module.exit_json(changed=False)
```
Comment (remixtj) [inline lib/ansible/modules/storage/emc/emc_vplex_storagevolume.py]: When hitting an error you must exit with module.fail_json

File: lib/ansible/modules/storage/emc/emc_vplex_storagevolume.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+    # Check provided storage-volume name exists
+    if volume_name in storage_volumes:
+        logger.error('Name duplication occured. Provided storage-volume name have already configured on VPLEX.')
+        logger.error('Please specify another storage-volume name in group_vars/all.yml and retry playbooks.')
+        module.exit_json(changed=False)
```
Comment (remixtj) [inline lib/ansible/modules/storage/emc/emc_vplex_storagevolume.py]: When hitting an error you must exit with module.fail_json

File: lib/ansible/modules/storage/emc/emc_vplex_storagevolume.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    else:
+        try:
+            claim_storage_volume(vplex=vplex, volume_name=volume_name, vpd_id=vpd_id)
+        except Exception:
+            logger.error('Failed to claim storage-volumes. Retry playbook after checking backend volumes and masking on FC-Switches.')
+            module.exit_json(changed=False)
```
Comment (remixtj) [inline lib/ansible/modules/storage/emc/emc_vplex_storagevolume.py]: When hitting an error you must exit with module.fail_json

File: lib/ansible/modules/storage/emc/emc_vplex_storagevolume.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    claim_uri = '/vplex/storage-volume+claim'
+    claiming_data = '{\"args\":\"-n ' + volume_name + ' -d ' + vpd_id + ' -f\"}'
+    response = vplex.https_post(urlsuffix=claim_uri, data=claiming_data)
+    return response
+
+
+def main():
+    log_directory = "./logs"
+    if not os.path.exists(log_directory):
+        os.makedirs(log_directory)
```
Comment (remixtj) [inline lib/ansible/modules/storage/emc/emc_vplex_storagevolume.py:136]: I may not be allowed to write to current directory. Maybe log path should be passed as parameter.

**Comment by hwakabh** [comment]
@remixtj Thanks to taking your time and reviewing codes. As pointed out, modified the module's behavior when it hit some errors, and made codes allowed user to provide path of this module logs.

**Comment by remixtj** [comment]
other notes: the module is not idempotent, because if the storage-volume, with the given setup is already present it fails. Instead, you should check if the volume you're trying to create already exists with the configuration you put as input. Additionally, there's no state argument, so user cannot remove a storage volume (with state: absent as example)

**Comment by hwakabh** [comment]
Closed for retest.

**Comment by hwakabh** [comment]
Reopened for retesting by CI.

**Comment by remixtj** [comment]
Seems good now but i didn't tested because i have no VPLEX devices to test against now.

**Comment by hwakabh** [comment]
@remixtj 
Thanks to your comments, and reviewing.

As my understanding, source codes for new ansible modules in new contributor's PRs would be tested with both CI-testings(by Shippable) and existing contributor's reviews, and these are not including unit-testings.

Considering the resources which seems to be vendor-specific like this PR, how do the new source codes been merged to master branch ??
Is it enough to note here the evidences, such as logs, of unit-testings in private environment ??

**Pull request closed without merging**